src/GafferScene/IECoreScenePreview/PrimitiveAlgo.cpp

```diff
@@ -0,0 +1,1188 @@
+		if( !vecVar && !vecConstVar )
+		{
+			continue;
+		}
+
+		GeometricData::Interpretation interp = vecVar ? vecVar->getInterpretation() : vecConstVar->getInterpretation();
+		if( !(
+			interp == GeometricData::Interpretation::Point ||
+			interp == GeometricData::Interpretation::Vector ||
+			interp == GeometricData::Interpretation::Normal
+		) )
+		{
+			continue;
+		}
+
+		if( vecVar )
+		{
+			for( Imath::V3f &i : vecVar->writable() )
+			{
+				Canceller::check( canceller );
+				transformPrimVarValue( i, i, matrix, normalMatrix, interp );
+			}
+		}
+		else
+		{
+			// Fairly weird corner case, but technically Constant primvars could need transforming too
+			transformPrimVarValue( vecConstVar->writable(), vecConstVar->writable(), matrix, normalMatrix, interp );
+		}
+	}
+}
+
+IECoreScene::PrimitivePtr PrimitiveAlgo::mergePrimitives(
+	const std::vector< std::pair< const IECoreScene::Primitive*, Imath::M44f > > &primitives,
+	const IECore::Canceller *canceller
+)
+{
+	IECoreScene::TypeId resultTypeId = (IECoreScene::TypeId)IECore::InvalidTypeId;
+
+	// Mesh specific globals
+	std::string resultMeshInterpolation;
+
+	// Curve specific globals
+	const CubicBasisf invalidBasis( Imath::M44f( 0.0f ), 0 );
+	CubicBasisf resultCurvesBasis( invalidBasis );
+	bool resultCurvesPeriodic( false );
+
+	// Data we need to store for each primvar we output
+	struct PrimVarInfo
+	{
+		PrimVarInfo( IECore::TypeId t, GeometricData::Interpretation interpretation, int numMeshes )
+			: invalid( false ), needsVerts( false ), needsFaces( false ), needsCurveKnots( false ),
+			typeId( t ), interpret( interpretation ), interpretInvalid( false ), indexed( false ),
+			numData( numMeshes, 0 )
+		{
+		}
+		// Primitive variables that are invalid should be reported once when we first realize it's invalid,
+		// and subsequently ignored.
+		bool invalid;
+
+		// These flags determine what interpolation is necessary for the result
+		bool needsVerts;
+		bool needsFaces;
+		bool needsCurveKnots;
+
+		// Computed from the flags above.
+		PrimitiveVariable::Interpolation interpolation;
+
+		// Need to track typeId so we can make sure everything matches
+		IECore::TypeId typeId;
+
+		GeometricData::Interpretation interpret;
+		bool interpretInvalid;
+
+		// The only case where we don't index the output is if all the input interpolations match, and none
+		// of the inputs are indexed - hopefully this is the common case though.
+		bool indexed;
+
+		// We need to collect the data size before we can allocate the output primvars
+		std::vector<unsigned int> numData;
+		std::vector<unsigned int> accumDataSizes;
+	};
+
+	std::unordered_map< IECore::InternedString, PrimVarInfo > varInfos;
+
+	//
+	// Before we can even start counting the sizes of things, we need to gather information about what
+	// kinds of primitives and primvars we're dealing with.
+	//
+
+	for( unsigned int i = 0; i < primitives.size(); i++ )
```

Comment:
Maybe range-for, so I don't have to remember that `i` indexes `primitives` all the way to line 726?

Reply:
Addressed as part of overall refactor.